File: bindings/java/src/main/java/com/strling/core/STRlingParseError.java
package com.strling.core;

import java.util.HashMap;
import java.util.Map;

/**
 * STRling Error Classes - Rich Error Handling for Instructional Diagnostics
 *
 * <p>This module provides enhanced error classes that deliver context-aware,
 * instructional error messages. The STRlingParseError class stores detailed
 * information about syntax errors including position, context, and beginner-friendly
 * hints for resolution.</p>
 */
public class STRlingParseError extends RuntimeException {
    /**
     * A concise description of what went wrong.
     */
    private final String message;
    
    /**
     * The character position (0-indexed) where the error occurred.
     */
    private final int pos;
    
    /**
     * The full input text being parsed.
     */
    private final String text;
    
    /**
     * An instructional hint explaining how to fix the error.
     */
    private final String hint;
    
    /**
     * Initializes a STRlingParseError.
     * 
     * @param message A concise description of what went wrong
     * @param pos The character position (0-indexed) where the error occurred
     * @param text The full input text being parsed (default: "")
     * @param hint An instructional hint explaining how to fix the error (default: null)
     */
    public STRlingParseError(String message, int pos, String text, String hint) {
        super(formatError(message, pos, text, hint));
        this.message = message;
        this.pos = pos;
        this.text = text != null ? text : "";
        this.hint = hint;
    }
    
    /**
     * Initializes a STRlingParseError without a hint.
     * 
     * @param message A concise description of what went wrong
     * @param pos The character position (0-indexed) where the error occurred
     * @param text The full input text being parsed
     */
    public STRlingParseError(String message, int pos, String text) {
        this(message, pos, text, null);
    }
    
    /**
     * Initializes a STRlingParseError without text or hint.
     * 
     * @param message A concise description of what went wrong
     * @param pos The character position (0-indexed) where the error occurred
     */
    public STRlingParseError(String message, int pos) {
        this(message, pos, "", null);
    }
    
    /**
     * Gets the error message.
     * 
     * @return The error message
     */
    public String getErrorMessage() {
        return message;
    }
    
    /**
     * Gets the error position.
     * 
     * @return The character position where the error occurred
     */
    public int getPos() {
        return pos;
    }
    
    /**
     * Gets the input text.
     * 
     * @return The full input text being parsed
     */
    public String getText() {
        return text;
    }
    
    /**
     * Gets the hint.
     * 
     * @return The instructional hint, or null if none provided
     */
    public String getHint() {
        return hint;
    }
    
    /**
     * Formats the error in the visionary state format.
     * 
     * @param message The error message
     * @param pos The character position where the error occurred
     * @param text The full input text
     * @param hint An optional hint
     * @return A formatted error message with context and hints
     */
    private static String formatError(String message, int pos, String text, String hint) {
        if (text == null || text.isEmpty()) {
            // Fallback to simple format if no text provided
            return message + " at position " + pos;
        }
        
        // Find the line containing the error
        String[] lines = text.split("\n", -1);
        int currentPos = 0;
        int lineNum = 1;
        String lineText = "";
        int col = pos;
        
        boolean found = false;
        for (int i = 0; i < lines.length; i++) {
            String line = lines[i];
            int lineLen = line.length() + 1; // +1 for newline
            if (currentPos + lineLen > pos) {
                lineNum = i + 1;
                lineText = line;
                col = pos - currentPos;
                found = true;
                break;
            }
            currentPos += lineLen;
        }
        
        if (!found) {
            // Error is beyond the last line
            if (lines.length > 0) {
                lineNum = lines.length;
                lineText = lines[lines.length - 1];
                col = lineText.length();
            } else {
                lineText = text;
                col = pos;
            }
        }
        
        // Build the formatted error message
        StringBuilder sb = new StringBuilder();
        sb.append("STRling Parse Error: ").append(message).append("\n");
        sb.append("\n");
        sb.append("> ").append(lineNum).append(" | ").append(lineText).append("\n");
        sb.append(">   | ");
        for (int i = 0; i < col; i++) {
            sb.append(" ");
        }
        sb.append("^\n");
        
        if (hint != null && !hint.isEmpty()) {
            sb.append("\n");
            sb.append("Hint: ").append(hint);
        }
        
        return sb.toString();
    }
    
    /**
     * Gets the formatted error message.
     * 
     * @return The formatted error string
     */
    public String toFormattedString() {
        return getMessage();
    }
    
    /**
     * Converts the error to LSP Diagnostic format.
     * 
     * <p>Returns a Map compatible with the Language Server Protocol
     * Diagnostic specification, which can be serialized to JSON for
     * communication with LSP clients.</p>
     * 
     * @return A Map containing:
     *         <ul>
     *         <li>range: The line/column range where the error occurred</li>
     *         <li>severity: Error severity (1 = Error)</li>
     *         <li>message: The error message with hint if available</li>
     *         <li>source: "STRling"</li>
     *         <li>code: A normalized error code derived from the message</li>
     *         </ul>
     */
    public Map<String, Object> toLspDiagnostic() {
        // Find the line and column containing the error
        String[] lines = text.isEmpty() ? new String[0] : text.split("\n", -1);
        int currentPos = 0;
        int lineNum = 0; // 0-indexed for LSP
        int col = pos;
        
        boolean found = false;
        for (int i = 0; i < lines.length; i++) {
            String line = lines[i];
            int lineLen = line.length() + 1; // +1 for newline
            if (currentPos + lineLen > pos) {
                lineNum = i;
                col = pos - currentPos;
                found = true;
                break;
            }
            currentPos += lineLen;
        }
        
        if (!found) {
            // Error is beyond the last line
            if (lines.length > 0) {
                lineNum = lines.length - 1;
                col = lines[lines.length - 1].length();
            } else {
                lineNum = 0;
                col = pos;
            }
        }
        
        // Build the diagnostic message
        String diagnosticMessage = message;
        if (hint != null && !hint.isEmpty()) {
            diagnosticMessage += "\n\nHint: " + hint;
        }
        
        // Create error code from message (normalize to snake_case)
        String errorCode = message.toLowerCase();
        String[] charsToReplace = {" ", "'", "\"", "(", ")", "[", "]", "{", "}", "\\", "/"};
        for (String ch : charsToReplace) {
            errorCode = errorCode.replace(ch, "_");
        }
        // Remove multiple underscores and empty parts
        errorCode = errorCode.replaceAll("_+", "_").replaceAll("^_|_$", "");
        
        // Build range
        Map<String, Integer> start = new HashMap<>();
        start.put("line", lineNum);
        start.put("character", col);
        
        Map<String, Integer> end = new HashMap<>();
        end.put("line", lineNum);
        end.put("character", col + 1);
        
        Map<String, Object> range = new HashMap<>();
        range.put("start", start);
        range.put("end", end);
        
        // Build diagnostic
        Map<String, Object> diagnostic = new HashMap<>();
        diagnostic.put("range", range);
        diagnostic.put("severity", 1); // 1 = Error, 2 = Warning, 3 = Information, 4 = Hint
        diagnostic.put("message", diagnosticMessage);
        diagnostic.put("source", "STRling");
        diagnostic.put("code", errorCode);
        
        return diagnostic;
    }
}

File: bindings/csharp/src/STRling/Core/STRlingParseError.cs
namespace Strling.Core;

/// <summary>
/// STRling Error Classes - Rich Error Handling for Instructional Diagnostics
/// 
/// This module provides enhanced error classes that deliver context-aware,
/// instructional error messages. The STRlingParseError class stores detailed
/// information about syntax errors including position, context, and beginner-friendly
/// hints for resolution.
/// </summary>

using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// Rich parse error with position tracking and instructional hints.
/// 
/// This error class transforms parse failures into learning opportunities by
/// providing:
/// - The specific error message
/// - The exact position where the error occurred
/// - The full line of text containing the error
/// - A beginner-friendly hint explaining how to fix the issue
/// </summary>
public class STRlingParseError : Exception
{
    /// <summary>
    /// A concise description of what went wrong.
    /// </summary>
    public string ErrorMessage { get; }
    
    /// <summary>
    /// The character position (0-indexed) where the error occurred.
    /// </summary>
    public int Pos { get; }
    
    /// <summary>
    /// The full input text being parsed.
    /// </summary>
    public string Text { get; }
    
    /// <summary>
    /// An instructional hint explaining how to fix the error.
    /// </summary>
    public string? Hint { get; }

    /// <summary>
    /// Initialize a STRlingParseError.
    /// </summary>
    /// <param name="message">A concise description of what went wrong</param>
    /// <param name="pos">The character position (0-indexed) where the error occurred</param>
    /// <param name="text">The full input text being parsed (default: "")</param>
    /// <param name="hint">An instructional hint explaining how to fix the error (default: null)</param>
    public STRlingParseError(
        string message,
        int pos,
        string text = "",
        string? hint = null)
        : base(FormatError(message, pos, text, hint))
    {
        ErrorMessage = message;
        Pos = pos;
        Text = text;
        Hint = hint;
    }

    /// <summary>
    /// Format the error in the visionary state format.
    /// </summary>
    /// <returns>A formatted error message with context and hints</returns>
    private static string FormatError(string message, int pos, string text, string? hint)
    {
        if (string.IsNullOrEmpty(text))
        {
            // Fallback to simple format if no text provided
            return $"{message} at position {pos}";
        }

        // Find the line containing the error
        var lines = text.Split('\n');
        var currentPos = 0;
        var lineNum = 1;
        var lineText = "";
        var col = pos;

        for (int i = 0; i < lines.Length; i++)
        {
            var line = lines[i];
            var lineLen = line.Length + 1; // +1 for newline
            if (currentPos + lineLen > pos)
            {
                lineNum = i + 1;
                lineText = line;
                col = pos - currentPos;
                break;
            }
            currentPos += lineLen;
        }

        // If error is beyond the last line
        if (string.IsNullOrEmpty(lineText))
        {
            if (lines.Length > 0)
            {
                lineNum = lines.Length;
                lineText = lines[^1];
                col = lineText.Length;
            }
            else
            {
                lineText = text;
                col = pos;
            }
        }

        // Build the formatted error message
        var parts = new List<string>
        {
            $"STRling Parse Error: {message}",
            "",
            $"> {lineNum} | {lineText}",
            $">   | {new string(' ', col)}^"
        };

        if (!string.IsNullOrEmpty(hint))
        {
            parts.Add("");
            parts.Add($"Hint: {hint}");
        }

        return string.Join("\n", parts);
    }

    /// <summary>
    /// Backwards/JS-friendly alias for getting the formatted error string.
    /// </summary>
    /// <returns>The formatted error message (same as ToString())</returns>
    public string ToFormattedString()
    {
        return Message;
    }

    /// <summary>
    /// Convert the error to LSP Diagnostic format.
    /// 
    /// Returns a dictionary compatible with the Language Server Protocol
    /// Diagnostic specification, which can be serialized to JSON for
    /// communication with LSP clients.
    /// </summary>
    /// <returns>
    /// A dictionary containing:
    /// - range: The line/column range where the error occurred
    /// - severity: Error severity (1 = Error)
    /// - message: The error message with hint if available
    /// - source: "STRling"
    /// - code: A normalized error code derived from the message
    /// </returns>
    public Dictionary<string, object> ToLspDiagnostic()
    {
        // Find the line and column containing the error
        var lines = !string.IsNullOrEmpty(Text) ? Text.Split('\n') : Array.Empty<string>();
        var currentPos = 0;
        var lineNum = 0; // 0-indexed for LSP
        var col = Pos;

        for (int i = 0; i < lines.Length; i++)
        {
            var line = lines[i];
            var lineLen = line.Length + 1; // +1 for newline
            if (currentPos + lineLen > Pos)
            {
                lineNum = i;
                col = Pos - currentPos;
                break;
            }
            currentPos += lineLen;
        }

        // If error is beyond the last line
        if (lineNum >= lines.Length && lines.Length > 0)
        {
            lineNum = lines.Length - 1;
            col = lines[^1].Length;
        }
        else if (lines.Length == 0)
        {
            lineNum = 0;
            col = Pos;
        }

        // Build the diagnostic message
        var diagnosticMessage = ErrorMessage;
        if (!string.IsNullOrEmpty(Hint))
        {
            diagnosticMessage += $"\n\nHint: {Hint}";
        }

        // Create error code from message (normalize to snake_case)
        var errorCode = ErrorMessage.ToLower();
        foreach (var ch in new[] { ' ', '\'', '"', '(', ')', '[', ']', '{', '}', '\\', '/' })
        {
            errorCode = errorCode.Replace(ch, '_');
        }
        errorCode = string.Join("_", errorCode.Split('_').Where(s => !string.IsNullOrEmpty(s)));

        return new Dictionary<string, object>
        {
            ["range"] = new Dictionary<string, object>
            {
                ["start"] = new Dictionary<string, object>
                {
                    ["line"] = lineNum,
                    ["character"] = col
                },
                ["end"] = new Dictionary<string, object>
                {
                    ["line"] = lineNum,
                    ["character"] = col + 1
                }
            },
            ["severity"] = 1, // 1 = Error, 2 = Warning, 3 = Information, 4 = Hint
            ["message"] = diagnosticMessage,
            ["source"] = "STRling",
            ["code"] = errorCode
        };
    }
}
